File: Information-Security-Lab/EXP2/playfair.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define GRID_SIZE 5
#define MAX_INPUT_SIZE 100

void create_grid(char key[], char grid[GRID_SIZE][GRID_SIZE]) {
    int alphabets[26] = {0};
    int key_len = strlen(key);
    int grid_index = 0;

    for (int i = 0; i < key_len; i++) {
        key[i] = tolower(key[i]);
        if (key[i] == 'j') {
            key[i] = 'i';
        }
        if (!alphabets[key[i] - 'a']) {
            alphabets[key[i] - 'a'] = 1;
        }
    }

    for (int i = 0; i < key_len; i++) {
        if (isalpha(key[i]) && alphabets[key[i] - 'a'] == 1) {
            grid[grid_index / GRID_SIZE][grid_index % GRID_SIZE] = key[i];
            alphabets[key[i] - 'a'] = 2;
            grid_index++;
        }
    }

    for (char ch = 'a'; ch <= 'z'; ch++) {
        if (ch == 'j') continue;
        if (alphabets[ch - 'a'] == 0) {
            grid[grid_index / GRID_SIZE][grid_index % GRID_SIZE] = ch;
            grid_index++;
        }
    }
}

void create_pairs(char plain_text[], char pairs[][2], int *num_pairs) {
    int text_len = strlen(plain_text);
    int index = 0;

    for (int i = 0; i < text_len; i++) {
        if (plain_text[i] == ' ') continue;
        if (index % 2 == 1 && plain_text[i] == pairs[index / 2][0]) {
            pairs[index / 2][index % 2] = 'x';
            index++;
        }
        pairs[index / 2][index % 2] = tolower(plain_text[i]);
        index++;
    }

    if (index % 2 == 1) {
        pairs[index / 2][1] = 'x';
        index++;
    }

    *num_pairs = index / 2;
}

void find_position(char letter, char grid[GRID_SIZE][GRID_SIZE], int *row, int *col) {
    for (int i = 0; i < GRID_SIZE; i++) {
        for (int j = 0; j < GRID_SIZE; j++) {
            if (grid[i][j] == letter) {
                *row = i;
                *col = j;
                return;
            }
        }
    }
}

void encrypt_pairs(char pairs[][2], int num_pairs, char grid[GRID_SIZE][GRID_SIZE], char cipher_text[]) {
    int index = 0;

    for (int i = 0; i < num_pairs; i++) {
        int r1, c1, r2, c2;
        find_position(pairs[i][0], grid, &r1, &c1);
        find_position(pairs[i][1], grid, &r2, &c2);

        if (r1 == r2) {
            cipher_text[index++] = grid[r1][(c1 + 1) % GRID_SIZE];
            cipher_text[index++] = grid[r2][(c2 + 1) % GRID_SIZE];
        } else if (c1 == c2) {
            cipher_text[index++] = grid[(r1 + 1) % GRID_SIZE][c1];
            cipher_text[index++] = grid[(r2 + 1) % GRID_SIZE][c2];
        } else {
            cipher_text[index++] = grid[r1][c2];
            cipher_text[index++] = grid[r2][c1];
        }
    }

    cipher_text[index] = '\0';
}

void decrypt_pairs(char cipher_text[], char grid[GRID_SIZE][GRID_SIZE], char plain_text[]) {
    int text_len = strlen(cipher_text);
    int index = 0;

    for (int i = 0; i < text_len; i += 2) {
        int r1, c1, r2, c2;
        find_position(cipher_text[i], grid, &r1, &c1);
        find_position(cipher_text[i + 1], grid, &r2, &c2);

        if (r1 == r2) {
            plain_text[index++] = grid[r1][(c1 - 1 + GRID_SIZE) % GRID_SIZE];
            plain_text[index++] = grid[r2][(c2 - 1 + GRID_SIZE) % GRID_SIZE];
        } else if (c1 == c2) {
            plain_text[index++] = grid[(r1 - 1 + GRID_SIZE) % GRID_SIZE][c1];
            plain_text[index++] = grid[(r2 - 1 + GRID_SIZE) % GRID_SIZE][c2];
        } else {
            plain_text[index++] = grid[r1][c2];
            plain_text[index++] = grid[r2][c1];
        }
    }

    plain_text[index] = '\0';
}

int main() {
    char key[MAX_INPUT_SIZE];
    char text[MAX_INPUT_SIZE];
    char grid[GRID_SIZE][GRID_SIZE];
    char pairs[50][2];
    int num_pairs;
    char choice;

    printf("Enter the key: ");
    fgets(key, MAX_INPUT_SIZE, stdin);
    key[strcspn(key, "\n")] = 0; // Remove trailing newline

    printf("Do you want to encrypt (e) or decrypt (d)? ");
    scanf("%c", &choice);
    getchar(); // Consume the newline character left in the buffer

    if (choice == 'e' || choice == 'E') {
        printf("Enter the plain text: ");
        fgets(text, MAX_INPUT_SIZE, stdin);
        text[strcspn(text, "\n")] = 0; // Remove trailing newline

        create_grid(key, grid);
        create_pairs(text, pairs, &num_pairs);

        char cipher_text[100];
        encrypt_pairs(pairs, num_pairs, grid, cipher_text);

        printf("Cipher Text:\n%s\n", cipher_text);
    } else if (choice == 'd' || choice == 'D') {
        printf("Enter the cipher text: ");
        fgets(text, MAX_INPUT_SIZE, stdin);
        text[strcspn(text, "\n")] = 0; // Remove trailing newline

        create_grid(key, grid);

        char plain_text[100];
        decrypt_pairs(text, grid, plain_text);

        printf("Decrypted Text:\n%s\n", plain_text);
    } else {
        printf("Invalid choice. Please enter 'e' for encryption or 'd' for decryption.\n");
    }

    return 0;
}

File: Information-Security-Lab/EXP2/playfair.py
key = 'Kathmandu'
plain_text = 'A quick brown fox jumps over little lazy dog'

def create_grid(key):
    '''Function to create the 5x5 playfair grid'''
    key = key.lower()
    grid = []
    alphabets = list(chr(x) for x in range(97, 123) if x != 'j')

    total_unique_letters = []

    for each_letter in key:
        if each_letter == 'j' or each_letter == 'i':
            if 'i' not in total_unique_letters:
                total_unique_letters.append('i')
        elif each_letter not in total_unique_letters and each_letter != ' ':        
            total_unique_letters.append(each_letter)

    for each_letter in alphabets:
        if each_letter not in total_unique_letters:
            total_unique_letters.append(each_letter)

    counter = 0
    size = 5
    while counter + size <= len(total_unique_letters):
        temp = total_unique_letters[counter:counter + size]
        counter += size
        grid.append(temp)

    return grid

def create_pair(plain_text):
    '''Function to form pairs of two letters in the plain text'''
    pairs = []
    letters = list(x for x in plain_text.lower() if x != ' ')
    i = 0
    
    while i < len(letters):
        if i == len(letters) - 1:  # If there's only one character left
            pairs.append([letters[i], 'x'])
            i += 1
        elif letters[i] != letters[i + 1]:
            pairs.append([letters[i], letters[i + 1]])
            i += 2
        else:
            pairs.append([letters[i], 'x'])
            i += 1

    return pairs

def find_position(letter, grid):
    '''Function to find the position of a letter in the grid'''
    for i in range(5):
        for j in range(5):
            if grid[i][j] == letter:
                return i, j
    return -1, -1

def encrypt(pairs, grid):
    '''Function to encrypt pairs using the Playfair cipher'''
    cipher_text = []

    for pair in pairs:
        r1, c1 = find_position(pair[0], grid)
        r2, c2 = find_position(pair[1], grid)
        
        if r1 == r2:  # Same row
            cipher_text.append(grid[r1][(c1 + 1) % 5])
            cipher_text.append(grid[r2][(c2 + 1) % 5])
        elif c1 == c2:  # Same column
            cipher_text.append(grid[(r1 + 1) % 5][c1])
            cipher_text.append(grid[(r2 + 1) % 5][c2])
        else:  # Rectangle
            cipher_text.append(grid[r1][c2])
            cipher_text.append(grid[r2][c1])

    return ''.join(cipher_text)

def decrypt(cipher_text, grid):
    '''Function to decrypt the cipher text using the Playfair cipher'''
    pairs = [list(cipher_text[i:i+2]) for i in range(0, len(cipher_text), 2)]
    plain_text = []

    for pair in pairs:
        r1, c1 = find_position(pair[0], grid)
        r2, c2 = find_position(pair[1], grid)
        
        if r1 == r2:  # Same row
            plain_text.append(grid[r1][(c1 - 1) % 5])
            plain_text.append(grid[r2][(c2 - 1) % 5])
        elif c1 == c2:  # Same column
            plain_text.append(grid[(r1 - 1) % 5][c1])
            plain_text.append(grid[(r2 - 1) % 5][c2])
        else:  # Rectangle
            plain_text.append(grid[r1][c2])
            plain_text.append(grid[r2][c1])

    return ''.join(plain_text)

def main():
    grid = create_grid(key)
    pairs = create_pair(plain_text)
    cipher_text = encrypt(pairs, grid)

    print("Cipher Text:")
    print(cipher_text)

    decrypted_text = decrypt(cipher_text, grid)
    print("\nDecrypted Text:")
    print(decrypted_text)
    
if __name__ == '__main__':
    main()
